plugins/goodix-fp/fu-goodixfp-device.c

```diff
@@ -0,0 +1,402 @@
+}
+
+static gchar *
+fu_goodixfp_device_get_version (FuGoodixFpDevice *self, GError **error)
+{
+  gxfp_cmd_response_t reponse = {0, };
+  guint8 dummy = 0;
+  gchar ver[9] = {0};
+
+  g_autoptr(GByteArray) request = g_byte_array_new ();
+  fu_byte_array_append_uint8 (request, dummy);
+  if (!fu_goodixfp_device_cmd_xfer (self, GX_CMD_VERSION, GX_CMD1_DEFAULT,
+                                    0,
+                                    request,
+                                    &reponse,
+                                    TRUE,
+                                    error))
+    return NULL;
+  memcpy (ver, reponse.version_info.fwversion, 8);
+  return g_strdup (ver);
+}
+
+static gboolean
+fu_goodixfp_device_update_init (FuGoodixFpDevice *self, GError **error)
+{
+  gxfp_cmd_response_t reponse = {0, };
+  g_autoptr(GByteArray) request = g_byte_array_new ();
+  /* update initial */
+  if (!fu_goodixfp_device_cmd_xfer (self, GX_CMD_UPGRADE, GX_CMD_UPGRADE_INIT,
+                                    0,
+                                    request,
+                                    &reponse,
+                                    TRUE,
+                                    error))
+    return FALSE;
+  return (reponse.result == 0) ? TRUE : FALSE;
+}
+
+static gboolean
+fu_goodixfp_device_attach (FuDevice *device, GError **error)
+{
+  FuGoodixFpDevice *self = FU_GOODIXFP_DEVICE (device);
+  g_autoptr(GByteArray) request = g_byte_array_new ();
+  gxfp_cmd_response_t reponse = {0, };
+  
+  fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+  
+  /* reset device */
+  if (!fu_goodixfp_device_cmd_xfer (self, GX_CMD_RESET, 0x03,
+                                    0,
+                                    request,
+                                    &reponse,
+                                    FALSE,
+                                    error))
+    return FALSE;
+  if (reponse.result != 0)
+    return FALSE;
+    
+  fu_device_add_flag (device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+  return TRUE;
+}
+
+static gboolean
+fu_goodixfp_device_open (FuUsbDevice *device, GError **error)
+{
+  GUsbDevice *usb_device = fu_usb_device_get_dev (device);
+
+  if (!g_usb_device_claim_interface (usb_device, GX_USB_INTERFACE,
+                                     G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
+                                     error))
+    return FALSE;
+
+  /* success */
+  return TRUE;
+}
+
+static gboolean
+fu_goodixfp_device_setup (FuDevice *device, GError **error)
+{
+  FuGoodixFpDevice *self = FU_GOODIXFP_DEVICE (device);
+  g_autofree gchar *version = NULL;
+  g_autoptr(GError) error_local = NULL;
+
+  version = fu_goodixfp_device_get_version (self, &error_local);
+  if (version != NULL)
+    {
+      g_debug ("obtained fwver using API '%s'", version);
+      fu_device_set_version (device, version);
+    }
+  else
+    {
+      g_warning ("failed to get firmware version: %s",
+                 error_local->message);
+    }
```

Comment:
why would we fail to get the firmware version? surely in this case we should just return with FALSE from _setup() and not add the device?

Reply:
I agree